tools/config_explorer.cxx

```diff
@@ -0,0 +1,238 @@
+            << "  -dc              generate detailed config contents output\n"
+            << "  -a name          alternate application name\n"
+            << "  -v version       optional application version string\n"
+            << "  --prefix dir     optional non-standard install prefix directory\n"
+            << "\n"
+            << "If -ds is specified, the detailed search paths that apply to the application are\n"
+            << "displayed only otherwise, the config file is loaded.\n"
+            << "\n"
+            << "The option -dc only has effect when a config file is specified and causes a\n"
+            << "detailed output of the config entries.\n"
+            << "\n"
+            << "If -I or --path are specified, then the config file is only searched for using\n"
+            << "the specified path. The application name based paths are not used.\n"
+  ;
+
+  return;
+}
+
+
+// ------------------------------------------------------------------
+static int
+path_callback( const char*  argument,   // name of argument
+               const char*  value,      // value of argument
+               void*        call_data ) // data from register call
+{
+  const std::string p( value );
+
+  opt_path.push_back( p );
+  return 1;   // return true for OK
+}
+
+
+// ============================================================================
+config_explorer::
+config_explorer()
+{ }
+
+
+// ----------------------------------------------------------------------------
+int
+config_explorer::
+run( int argc, const char* argv[] )
+{
+  opt_app_name = applet_name();
+
+  kwiversys::CommandLineArguments arg;
+
+  arg.Initialize( argc, argv );
+  arg.StoreUnusedArguments( true );
+
+  arg.AddArgument( "-h",        argT::NO_ARGUMENT, &opt_help, "Display usage information" );
+  arg.AddArgument( "--help",    argT::NO_ARGUMENT, &opt_help, "Display usage information" );
+
+  // details
+  arg.AddArgument( "-ds",       argT::NO_ARGUMENT, &opt_detail_ds, "Display detailed application search path" );
+  arg.AddArgument( "-dc",       argT::NO_ARGUMENT, &opt_detail_dc, "Display detailed config contents" );
+
+  // manual search path
+  arg.AddCallback( "--path",    argT::SPACE_ARGUMENT, path_callback, 0, "Add directory to config search path" );
+  arg.AddCallback( "-I",        argT::CONCAT_ARGUMENT, path_callback, 0, "Add directory to config search path" );
+
+  // auto search path generation
+  arg.AddArgument( "-a",        argT::SPACE_ARGUMENT, &opt_app_name, "Application name" );
+  arg.AddArgument( "-v",        argT::SPACE_ARGUMENT, &opt_app_version, "Application version string" );
+  arg.AddArgument( "--prefix",  argT::SPACE_ARGUMENT, &opt_install_prefix,
+                   "Non-standard installation prefix. (e.g. /opt/kitware)" );
+
+  if ( ! arg.Parse() )
+  {
+    std::cerr << "Problem parsing arguments" << std::endl;
+    exit( 0 );
+  }
+
+  if ( opt_help )
+  {
+    usage( std::cout );
```

Comment:
should this be `usage( this->outstream() );`?

Reply:
I believe there is a standard for usage information, although I have not been able to find it, that specifies which output stream to use (either stdout or stderr). git uses stdout. The intent of the outstream() stream is to allow the output of an applet to be captured in a file.